### DISCUSSION TITLE
Issue: Large files continuously crash during data cleaning in QualCoder

### DISCUSSION THREAD

**Discussion by ChefDani** [ccbogel/QualCoder#938] [Q&A]
Hello,

I’ve been encountering an issue where large files continuously cause QualCoder to crash during the data cleaning process. These files seem to exceed the program's capacity to handle memory effectively, especially with extensive datasets and multiple coding processes. Here are the key details:

    QualCoder version: 3.5
    Operating system: Windows

    Steps to reproduce:
-         Load a large dataset into QualCoder ((original) file size over 3 MB).
-         Begin coding or cleaning the data.
-         After a few operations, the program crashes without saving the progress.

I’ve tried reducing the file size, but this is not feasible for my workflow. Is there a workaround for handling larger datasets or a solution that could help improve memory management?

Any assistance or guidance would be greatly appreciated.

Thanks!

**Comment by ccbogel** [comment]
Hmm, I don't have a large data file to test this. If it is possible to view yours - you can email me a zip of the project folder, I can test and see what could be done (if anything).
What do you mean by a large dataset ? - lots of files, or very large files. Specifically - very large text files.
Or also lots of attributes (e.g. case or file variables - are being loaded from a survey)?
When you state data cleaning - what are you doing for that exactly - is this editing a large text document before coding?

If it's a very large text file - can you edit it, i.e. clean it with another software before importing into QualCoder.
When opening the large file in Code Text: You can right-click on the file name and chunk text. So that smaller sections are loaded at a time.
From the wiki:
Large text files slow QualCoder. When opening a large text file (greater than 50,000 characters), right click on the file name. A menu option allows you to get the first approximately 50,000 characters, subsequent clicks will open the next approximately 50,000 and so on. Approximately, because QualCoder looks for a line-ending (such as a paragraph or sentence ending) to more nicely section the start and end of text chunks.

**Reply by ChefDani** [reply]
Sure thing, could you share your email-address for me to share my projectfolder.
My dataset has roughly 55 files upwards to 8 MB per file, furthermore doesnt it have a lot attributes.
With data cleaning, I indeed mean editing a large text document before coding.

Looking forward to your response!

In fortitude,
Dani Smith

**Reply by ccbogel** [reply]
As it is large, you will have to send it using: 
https://wetransfer.com/
my email is ccbogel at hotmail.com

I have some ideas. that involve chunking the data.

**Reply by kaixxx** [reply]
If you say 8 MB per file - do you mean plain text? That would be roughly 2500 book pages (assuming UTF-8 encoding) in a single file! 
I doubt that QualCoder (or any QDA software for that matter) is the right tool to analyze such massive amounts of text. But I may have misunderstood what you are trying to achieve.

**Reply by ccbogel** [reply]
Yes it is a huge amount of information. Gedit in Ubuntu or other Linux would likely manage it. I am looking at replacing the QTextEdit with QPlainTextEdit in the Mange File > View/Edit file. As this widget is better at handling large text file. But I am currently running into a couple of issue.

**Reply by ChefDani** [reply]
Regarding the 8 MB per file, that is by far the largest file and I took the PDF version in my zipped map in consideration. That report is 173 pages long (unedited), currently I narrowed it down to roughly 85 pages.

BTW I have sent the project through WeTransfer

Thanks in advance!

**Reply by ccbogel** [reply]
OK thanks. I have made one update so far, that will use the QPlainTextEdit which may be faster, and may not crash.

**Reply by kaixxx** [reply]
@ccbogel I'm also wondering if SQLite is able to handle such large blobs of text.

**Reply by kaixxx** [reply]
> roughly 85 pages

That sounds much more reasonable!

**Reply by ccbogel** [reply]
@kaixx, a good question. I checked. Sqlite can handle text up to 1 billion bytes. So character encoding (depending on encoding) uses several bytes, UTF-8 uses up to 4 bytes. So a max length string would be 250 million characters. Displaying that in the Qt Widgets would be the issue with memory limits (crashing) and slowness.

**Reply by ccbogel** [reply]
I am backing up my data before I open the zip file. So I may or may not be able to view it tonight. 10.35pm here.

**Comment by ccbogel** [comment]
OK I did one file update in the latest code (3.6) which might help a little. 

The main thing to make the editing faster is to do this:
Go to Mange Cases. Open that specific case in the Files Colum Then REMOVE the file from the case. 
Then go to Manage Files. Then you can edit the file text with speed.
Then when finished editing (click OK), re-assign that file to the Case.

The problem is that whenever a character is changed (added or deleted), QualCoder is using the difflib module, which is slow for large files.

**Reply by ChefDani** [reply]
Will definitely check!

**Reply by ccbogel** [reply]
OK I have identified why it is so slow. There is a module called difflib that I use in  DialogEditTextFile.update_positions()
difflib looks at differences in the previous to the changed text, when every character is added or deleted. This module is slow with large text files. There aremodules called diff_lib_patch and cdifflib which are faster. So I will look into this.

**Comment by ccbogel** [comment]
OK I have updated the code to use diff-lib-patch module instead of the difflib module. This makes editing text 20x faster and no visible slowness.
TODO need to update code_text edit text functionality to match.

**Reply by ChefDani** [reply]
How does one update the code_text edit text functionality, exactly on Qualcoder?

**Reply by ccbogel** [reply]
I will do that soon. but first there is a big branch merge to sort out.
Did you test out the edit speed change in Manage File > Edit text file from the latest code?
